walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,124 @@
+    }
+
+    Collection<String> mandatoryAttendees = request.getAttendees();
+    Collection<String> optionalAttendees = request.getOptionalAttendees();
+
+    // If there are no mandatory attendees, use only the optional attendees.
+    if (mandatoryAttendees.isEmpty()) {
+      return queryHelper(events, optionalAttendees, meetingDuration);
+    }
+
+    Collection<String> mandAndOpt = new HashSet<String>(mandatoryAttendees);
+    mandAndOpt.addAll(optionalAttendees);
+
+    Collection<TimeRange> timesForAll = queryHelper(events, mandAndOpt, meetingDuration);
+    // If there are one or more time slots available for all potential attendees, return them.
+    if (timesForAll.size() > 0) {
+      return timesForAll;
+    }
+    else {
+      return queryHelper(events, mandatoryAttendees, meetingDuration);
+    }
+}
+
+/**
+ * Contains the bulk of the meeting query algorithm.
+ */
+public Collection<TimeRange> queryHelper(Collection<Event> events, Collection<String> attendees, long duration) {
+    ArrayList<TimeRange> takenTimes = new ArrayList<TimeRange>();   // A list of meeting times that are already taken.
+    ArrayList<TimeRange> validTimes = new ArrayList<TimeRange>();   // A list of valid times to meet for all attendees.
+
+    // Populate the taken meeting times list using the events param.
+    // If no one in the planned event was a relevant attendee, don't add the time range to taken times.
+    for (Event event: events) {
+      if (!Collections.disjoint(attendees, event.getAttendees())) {
+        takenTimes.add(event.getWhen());
+      }
+    }
+
+    // If no times are taken at all, return the full day.
+    if (takenTimes.isEmpty()) {
+      validTimes.add(TimeRange.WHOLE_DAY);
+      return validTimes;
+    }
+
+    // Sort the events by their start times and add the range from the beginning of the day to
+    // the start of the first meeting if enough time exists.
+    Collections.sort(takenTimes, TimeRange.ORDER_BY_START);
+    int firstEventStart = takenTimes.get(0).start();
+    if (firstEventStart >= duration) {
+      validTimes.add(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, takenTimes.get(0).start(), false));
+    }
+    
+    // Loop through the meetings (except the last) and add time ranges where no meetings exist.
+    int currEndTime = takenTimes.get(0).end();   // holds the most relevant end time. 
+    int currEventIndex = 0;                      // index of the event being compared in the loop--may not correspond to currEndTime.
+    while (currEventIndex < (takenTimes.size() - 1)) {
+      TimeRange currEvent = takenTimes.get(currEventIndex);
+      TimeRange nextEvent = takenTimes.get(currEventIndex + 1);
+
+      // If the current event overlaps the next event, it either contains the entire next event or its start.
+      // Therefore, update only the end time as needed.
+      if (currEvent.overlaps(nextEvent)) {
+        if (currEvent.contains(nextEvent)) {
+          currEndTime = currEvent.end();
```

Comment:
What if you have a meeting fully contained in another meeting, which is then in turn fully contained in a third meeting? Say you have meeting from 0-5, 1-4, 2-3.  I think this breaks in the second iteration of the loop (when currEvent = 1-4 and nextEvent =2-3). You'd update currEndTime to 4 when it should stay at 5.